docs/project-manager.1
.Dd January 1, 1980
.Dt project-manager 1
.Os Project Manager
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" enable line breaks after slashes
.cflags 4 /
.Sh NAME
.Nm project-manager
.Nd reconfigure a user environment
.
.
.
.Sh SYNOPSIS
.Nm project-manager
.Bro
.Cm build
.Cm | init Op Fl -switch Ar dir
.Cm | kitchen-sink \&...
.Cm | ks \&...
.Cm | instantiate
.Cm | edit
.Cm | expire-generations Ar timestamp
.Cm | generations
.Cm | help
.Cm | news
.Cm | option Ar option.name
.Cm | packages
.Cm | format \&...
.Cm | fmt \&...
.Cm | remove-generations Ar ID \&...
.Cm | uninstall
.Brc
.Op Fl A Ar attrPath
.Op Fl I Ar path
.Op Fl -flake Ar flake-uri
.Op Fl b Ar ext
.Op Bro Fl f | Fl -file Brc Ar path
.Op Bro Fl h | Fl -help Brc
.Op Fl -version
.Op Bro Fl n | Fl -dry-run Brc
.Op Fl -option Ar name Ar value
.Op Fl -cores Ar number
.Op Bro Fl j | Fl -max-jobs Brc Ar number
.Op Fl -option
.Op Fl -impure
.Op Fl -keep-failed
.Op Fl -keep-going
.Op Bro Fl L | Fl -print-build-logs Brc
.Op Fl -show-trace
.Op Fl -(no-)substitute
.Op Fl -no-out-link
.Op Fl -refresh
.Op Bro Fl v | Fl -verbose Brc
.
.Sh DESCRIPTION
.Pp
This command updates the user environment so that it corresponds to the configuration specified in
$PROJECT_ROOT/flake.nix.
.Pp
All operations using this tool expects a sub-command that indicates the operation to perform. It must be one of
.Pp
.Bl -tag -width Ds

.It Cm build
.RS 4
Build configuration into a result directory.
.RE

.It Cm init Op Fl -switch Op Ar dir
.RS 14
Generates an initial \&.config/project/default.nix file for the current user. If Nix flakes are
enabled, then this command also generates a flake.nix file.
.sp
If a path
.Ar dir
is given then the configuration will be generated in that directory. Otherwise, the configuration will be generated in
$PROJECT_ROOT/.config/project. The output directory will be created if it does not exist.
.sp
If the
.Fl -switch
option is given, then the generated configuration is activated.
.sp
Note, this command will not overwrite any existing files. It is therefore safe to initialize a configuration, edit it, and then re-run the
.Cm init
command with
.Fl -switch
enabled to activate the configuration.
.RE
.Pp

.It Cm instantiate
.RS 15
Instantiate the configuration and print the resulting derivation\&.
.RE
.Pp

.It Cm edit
.RS 16
Open the project configuration using the editor indicated by \fBEDITOR\fR\&.
.RE
.Pp

.It Cm expire-generations Ar timestamp
.RS 4
Remove generations older than
.Ar timestamp
where
.Ar timestamp
is interpreted as in the
.Fl d
argument of the
\fBdate\fR(1)
tool. For example
-30 days or 2018-01-01.
.RE
.PP

.It Cm generations
.RS 4
List all project environment generations\&.
.RE
.Pp

.It Cm help
.RS 4
Print tool help.
.RE
.Pp

.It Cm news
.RS 4
Show news entries in a pager.
.RE
.PP

.It Cm option Ar option.name
.RS 4
Inspect the given option name in the project configuration, like
\fBnixos-option\fR(8)\&.
.RE
.Pp

.It Cm packages
.RS 4
List all packages installed in project-manager-path.
.RE
.Pp

.It Cm format
.It Cm fmt
.RS 4
Run the configured formatter. This is usually much faster than running
\fBnix fmt\fR because it avoids evaluating the flake.
.RE
.Pp

.It Cm kitchen-sink
.It Cm ks
.RS 4
Run both \fBswitch\fR and \fBformat\fR commands.
.RE
.Pp

.It Cm remove-generations Ar ID \&...
.RS 4
Remove indicated generations. Use the
.Cm generations
sub-command to find suitable generation numbers.
.RE
.Pp

.It Cm switch
.RS 4
Build and activate the configuration\&.
.RE
.Pp

.It Cm uninstall
.RS 4
Remove Project Manager from the project environment\&. This will
.sp
.RE
.RS 4
.Bl -bullet
.It
remove all managed files from the project directory,
.RE
.sp
.RS 4
.It
remove packages installed through Project Manager from the project profile, and
.RE
.sp
.RS 4
.It
remove all Project Manager generations and make them available for immediate garbage collection\&.
.RE
.El
.sp
.RE
.El
.
.Sh OPTIONS
.Pp
The tool accepts the options
.Pp
.Bl -tag -width Ds
.It Cm Fl b Ar extension
.RS 4
Enable automatic resolution of collisions between unmanaged and managed files\&. The name of the original file will be suffixed by the given extension\&. For example,
.sp
.if n \{\
.RS 4
.\}
.nf
$ \fBproject\-manager \-b bck switch\fR
.fi
.if n \{\
.RE
.\}
.sp
will cause a colliding file
~/\&.config/foo\&.conf
to be moved to
~/\&.config/foo\&.conf\&.bck\&.
.RE
.Pp

.It Cm Fl f Ar path, Fl -file Ar path
.RS 4
Indicates the path to the Project Manager configuration file. If not given,
$PROJECT_ROOT/\&.config/project/default.nix
is used.
.RE
.Pp
.It Cm Fl h, Fl -help
.RS 4
Prints usage information for the
\fBproject\-manager\fR
tool.
.RE
.Pp

.It Cm Fl -version
.RS 4
Prints the version number of the
\fBproject\-manager\fR
tool.
.RE
.Pp
.It Cm Fl n, Fl -dry-run
.RS 4
Perform a dry-run of the given operation, only prints what actions would be taken.
.RE
.Pp

.It Cm Fl -option Ar name Ar value
.RS 4
Passed on to
\fBnix-build\fR(1)\&.
.RE
.Pp

.It Cm Fl -cores Ar number
.RS 4
Passed on to
\fBnix-build\fR(1)\&.
.RE
.Pp

.It Cm Fl j Ar number, Fl -max-jobs Ar number
.RS 4
Passed on to
\fBnix-build\fR(1)\&.
.RE
.\" TODO
.Pp
.It Cm Fl -debug
.RS 4
Passed on to
\fBnix-build\fR(1)\&.
.RE
.Pp
.It Cm Fl -impure
.RS 4
Passed on to
\fBnix-build\fR(1)\&.
.RE
.Pp

.It Cm Fl -keep-failed
.RS 4
Passed on to
\fBnix-build\fR(1)\&.
.RE
.Pp

.It Cm Fl -keep-going
.RS 4
Passed on to
\fBnix-build\fR(1)\&.
.RE
.Pp

.It Cm Fl L, Fl -print-build-logs
.RS 4
Passed on to
\fBnix build\fR()
when building from a flake\&.
.RE
.Pp

.It Cm Fl -show-trace
.RS 4
Passed on to
\fBnix-build\fR(1)\&.
.RE
.Pp

.It Cm Fl -(no-)substitute
.RS 4
Passed on to
\fBnix-build\fR(1)\&.
.RE
.Pp

.It Cm Fl -no-out-link
.RS 4
Passed on to
\fBnix-build\fR(1)
when running
\fBproject\-manager build\fR\&.
.RE
.Pp

.It Cm Fl -refresh
.RS 4
Passed on to
\fBnix-build\fR(1)
.RE
.Pp

.It Cm Fl v, Fl -verbose
.RS 4
Activates verbose output\&.
.RE
.El

.Sh FILES
.Pp
$PROJECT_ROOT/.local/state/project\-manager/news\-read\-ids
.RS 4
Identifiers of news items that have been shown\&. Can be deleted to reset the read news indicator\&.
.RE

.Sh BUGS
.Pp
Please report any bugs on the
\m[blue]\fBproject issue tracker\fR\m[]\&.

.Sh SEE ALSO
.Pp
\fBproject-configuration.nix\fR(5)

.Sh AUTHOR
.Pp
\fBProject Manager contributors\fR
.RS 4
Author.
.RE

.Sh COPYRIGHT
.br
Copyright \(co 2017\(en2022 Home Manager contributors
.br


lib/bash/project-manager.bash
# This file contains a number of utilities for use by the project-manager tool and
# the generated Project Manager activation scripts. No guarantee is made about
# backwards or forward compatibility.

# Sets up colors suitable for the `errorEcho`, `warnEcho`, and `noteEcho`
# functions.
#
# The check for terminal output and color support is heavily inspired by
# https://unix.stackexchange.com/a/10065.
#
# The setup respects the `NO_COLOR` environment variable.
function setupColors() {
  normalColor=""
  errorColor=""
  warnColor=""
  noteColor=""

  # Enable colors for terminals, and allow opting out.
  if [[ ! -v NO_COLOR && -t 1 ]]; then
    # See if it supports colors.
    local ncolors
    ncolors=$(tput colors 2> /dev/null || echo 0)

    if [[ -n $ncolors && $ncolors -ge 8 ]]; then
      normalColor="$(tput sgr0)"
      errorColor="$(tput bold)$(tput setaf 1)"
      warnColor="$(tput setaf 3)"
      noteColor="$(tput bold)$(tput setaf 6)"
    fi
  fi
}

setupColors

function errorEcho() {
  echo "${errorColor}$*${normalColor}"
}

function warnEcho() {
  echo "${warnColor}$*${normalColor}"
}

function noteEcho() {
  echo "${noteColor}$*${normalColor}"
}

function _i() {
  local msgid="$1"
  shift

  # shellcheck disable=2059
  printf "$(gettext "$msgid")\n" "$@"
}

function _ip() {
  local msgid="$1"
  local msgidPlural="$2"
  local count="$3"
  shift 3

  # shellcheck disable=2059
  printf "$(ngettext "$msgid" "$msgidPlural" "$count")\n" "$@"
}

function _iError() {
  echo -n "${errorColor}"
  _i "$@"
  echo -n "${normalColor}"
}

function _iWarn() {
  echo -n "${warnColor}"
  _i "$@"
  echo -n "${normalColor}"
}

function _iNote() {
  echo -n "${noteColor}"
  _i "$@"
  echo -n "${normalColor}"
}

## Project Manager logic

function pm_listPackagesBySuffix() {
  # We attempt to use `--json` first (added in Nix 2.17). Otherwise
  # attempt to parse the legacy output format.
  {
    nix profile list --profile "$1" --json 2> /dev/null \
      | jq --raw-output --arg name "$2" '.elements[].storePaths[] | select(endswith($name))'
  } || {
    nix profile list --profile "$1" \
      | { grep "$2\$" || test $? = 1; } \
      | cut -d ' ' -f 4
  }
}

function pm_removePackagesBySuffix() {
  pm_listPackagesBySuffix "$1" "$2" \
    | xargs "${VERBOSE_ARG[@]}" "${DRY_RUN_CMD[@]}" --no-run-if-empty \
      nix profile remove "${VERBOSE_ARG[@]}" --profile "$1"
}

function setNixProfileCommands() {
  LIST_OUTPATH_CMD=('nix' 'profile' 'list')
  REMOVE_CMD='pm_removePackagesBySuffix'
}

function pm_setVerboseAndDryRun() {
  if [[ -v VERBOSE ]]; then
    export VERBOSE_ECHO='echo'
    export VERBOSE_ARG=('--verbose')
    export VERBOSE_RUN=()
  else
    export VERBOSE_ECHO='true'
    export VERBOSE_ARG=()
    export VERBOSE_RUN=('true')
  fi

  if [[ -v DRY_RUN ]]; then
    _i "This is a dry run"
    export DRY_RUN_CMD=('echo')
    export DRY_RUN_NULL="/dev/stdout"
  else
    "${VERBOSE_RUN[@]}" _i "This is a live run"
    export DRY_RUN_CMD=()
    export DRY_RUN_NULL=/dev/null
  fi
}

function setWorkDir() {
  if [[ ! -v WORK_DIR ]]; then
    WORK_DIR="$(mktemp --tmpdir -d project-manager-build.XXXXXXXXXX)"
    # shellcheck disable=2064
    trap "rm -r '$WORK_DIR'" EXIT
  fi
}

# Checks whether the 'flakes' and 'nix-command' Nix options are enabled.
function hasFlakeSupport() {
  type -p nix > /dev/null \
    && nix show-config 2> /dev/null \
    | grep experimental-features \
      | grep flakes \
      | grep -q nix-command
}

# Attempts to set the PROJECT_MANAGER_CONFIG global variable.
#
# If no configuration file can be found then this function will print
# an error message and exit with an error code.
function setConfigFile() {
  if [[ -v PROJECT_MANAGER_CONFIG ]]; then
    if [[ -e $PROJECT_MANAGER_CONFIG ]]; then
      PROJECT_MANAGER_CONFIG="$(realpath "$PROJECT_MANAGER_CONFIG")"
    else
      _i 'No configuration file found at %s' \
        "$PROJECT_MANAGER_CONFIG" >&2
      exit 1
    fi
  elif [[ ! -v PROJECT_MANAGER_CONFIG ]]; then
    local configHome="${PROJECT_ROOT}/.config"
    local pmConfigHome="$configHome/project"
    local defaultConfFile="$pmConfigHome/default.nix"
    local configFile

    if [[ -e $defaultConfFile ]]; then
      configFile="$defaultConfFile"
    fi

    if [[ -v configFile ]]; then
      PROJECT_MANAGER_CONFIG="$(realpath "$configFile")"
    else
      _i 'No configuration file found. Please create one at %s' \
        "$defaultConfFile" >&2
      exit 1
    fi
  fi
}

# Sets some useful Project Manager related paths as global read-only variables.
function setProjectManagerPathVariables() {
  # If called twice then just exit early.
  if [[ -v PM_DATA_HOME ]]; then
    return
  fi

  declare -r stateHome="$PROJECT_ROOT/.local/state"
  declare -r userNixStateDir="$stateHome/nix"

  declare -gr PM_DATA_HOME="$PROJECT_ROOT/.local/share/project-manager"
  declare -gr PM_STATE_DIR="$stateHome/project-manager"

  mkdir -p "$userNixStateDir/profiles"
  declare -gr PM_PROFILE_DIR="$userNixStateDir/profiles"
}

function setFlakeAttribute() {
  local flake="."
  local name
  name="$(nix eval --expr 'builtins.currentSystem' --impure --raw)"
  export FLAKE_CONFIG_URI="$flake#projectConfigurations.$name"
}

function pm_init() {
  # The directory where we should place the initial configuration.
  local confDir

  # Whether we should immediate activate the configuration.
  local switch

  # Whether we should create a flake file.
  local withFlake

  if hasFlakeSupport; then
    withFlake=1
  fi

  while (($# > 0)); do
    local opt="$1"
    shift

    case $opt in
      --no-flake)
        unset withFlake
        ;;
      --switch)
        switch=1
        ;;
      -*)
        _iError "%s: unknown option '%s'" "$0" "$opt" >&2
        exit 1
        ;;
      *)
        if [[ -v confDir ]]; then
          _i "Run '%s --help' for usage help" "$0" >&2
          exit 1
        else
          confDir="$opt"
        fi
        ;;
    esac
  done

  if [[ ! -v confDir ]]; then
    confDir="${PROJECT_ROOT}/.config/project"
  fi

  if [[ ! -e $confDir ]]; then
    mkdir -p "$confDir"
  fi

  if [[ ! -d $confDir ]]; then
    _iError "%s: unknown option '%s'" "$0" "$opt" >&2
    exit 1
  fi

  local confFile="$confDir/default.nix"
  local flakeFile="${PROJECT_ROOT}/flake.nix"

  if [[ -e $confFile ]]; then
    _i 'The file %s already exists, leaving it unchanged...' "$confFile"
  else
    _i 'Creating %s...' "$confFile"

    mkdir -p "$confDir"
    cp '@CONFIG_TEMPLATE@' "$confFile"
  fi

  if [[ ! -v withFlake ]]; then
    PROJECT_MANAGER_CONFIG="$confFile"
  else
    if [[ -e $flakeFile ]]; then
      _i 'The file %s already exists, leaving it unchanged...' "$flakeFile"
    else
      _i 'Creating %s...' "$flakeFile"

      mkdir -p "$confDir"
      cp "@FLAKE_TEMPLATE@" "$flakeFile"
    fi
  fi

  if [[ -v switch ]]; then
    echo
    _i "Creating initial Project Manager generation..."
    echo

    if pm_switch; then
      # translators: The "%s" specifier will be replaced by a file path.
      _i $'All done! The project-manager tool should now be installed and you can edit\n\n    %s\n\nto configure Project Manager. Run \'man project-configuration.nix\' to\nsee all available options.' \
        "$confFile"
      exit 0
    else
      # translators: The "%s" specifier will be replaced by a URL.
      _i $'Uh oh, the installation failed! Please create an issue at\n\n    %s\n\nif the error seems to be the fault of Project Manager.' \
        "https://github.com/nix-community/project-manager/issues"
      exit 1
    fi
  fi
}

function pm_buildFlake() {
  local extraArgs=("$@")

  if [[ -v VERBOSE ]]; then
    extraArgs=("${extraArgs[@]}" "--verbose")
  fi

  nix build \
    "${extraArgs[@]}" \
    "${PASSTHROUGH_OPTS[@]}"
}

# Presents news to the user as specified by the `news.display` option.
function presentNews() {
  local newsNixFile="$WORK_DIR/news.nix"
  pm_buildNews "$newsNixFile"

  local newsDisplay
  newsDisplay="$(nix-instantiate --eval --expr "(import ${newsNixFile}).meta.display" | xargs)"

  local newsNumUnread
  newsNumUnread="$(nix-instantiate --eval --expr "(import ${newsNixFile}).meta.numUnread" | xargs)"

  # shellcheck disable=2154
  if [[ $newsNumUnread -eq 0 ]]; then
    return
  elif [[ $newsDisplay == "silent" ]]; then
    return
  elif [[ $newsDisplay == "notify" ]]; then
    local cmd msg
    cmd="$(basename "$0")"
    msg="$(_ip \
      $'There is %d unread and relevant news item.\nRead it by running the command "%s news".' \
      $'There are %d unread and relevant news items.\nRead them by running the command "%s news".' \
      "$newsNumUnread" "$newsNumUnread" "$cmd")"

    # Not actually an error but here stdout is reserved for
    # nix-build output.
    echo $'\n'"$msg"$'\n' >&2

    if [[ -v DISPLAY ]] && type -P notify-send > /dev/null; then
      notify-send "Project Manager" "$msg" > /dev/null 2>&1 || true
    fi
  elif [[ $newsDisplay == "show" ]]; then
    pm_showNews --unread
  else
    _i 'Unknown "news.display" setting "%s".' "$newsDisplay" >&2
  fi
}

function pm_edit() {
  if [[ ! -v EDITOR || -z $EDITOR ]]; then
    # shellcheck disable=2016
    _i 'Please set the $EDITOR environment variable' >&2
    return 1
  fi

  setConfigFile

  # Don't quote $EDITOR in order to support values including options, e.g.,
  # "code --wait".
  #
  # shellcheck disable=2086
  exec $EDITOR "$PROJECT_MANAGER_CONFIG"
}

function pm_build() {
  if [[ ! -w . ]]; then
    _i 'Cannot run build in read-only directory' >&2
    return 1
  fi

  setWorkDir

  setFlakeAttribute
  pm_buildFlake \
    "$FLAKE_CONFIG_URI.packages.activation" \
    ${DRY_RUN+--dry-run} \
    ${NO_OUT_LINK+--no-link} \
    ${PRINT_BUILD_LOGS+--print-build-logs} \
    || return

  # presentNews
}

function pm_switch() {
  setFlakeAttribute
  nix run \
    "$FLAKE_CONFIG_URI.packages.activation" \
    ${PRINT_BUILD_LOGS+--print-build-logs} \
    ${VERBOSE+--verbose} \
    "${PASSTHROUGH_OPTS[@]}" \
    || return

  # presentNews
}

function pm_listGenerations() {
  setProjectManagerPathVariables

  # Whether to colorize the generations output.
  local color="never"
  if [[ ! -v NO_COLOR && -t 1 ]]; then
    color="always"
  fi

  pushd "$PM_PROFILE_DIR" > /dev/null || exit
  # shellcheck disable=2012
  ls --color=$color -gG --time-style=long-iso --sort time project-manager-*-link \
    | cut -d' ' -f 4- \
    | sed -E 's/project-manager-([[:digit:]]*)-link/: id \1/'
  popd > /dev/null || exit
}

# Removes linked generations. Takes as arguments identifiers of
# generations to remove.
function pm_removeGenerations() {
  setProjectManagerPathVariables
  pm_setVerboseAndDryRun

  pushd "$PM_PROFILE_DIR" > /dev/null || exit

  for generationId in "$@"; do
    local linkName="project-manager-$generationId-link"

    if [[ ! -e $linkName ]]; then
      _i 'No generation with ID %s' "$generationId" >&2
    elif [[ $linkName == $(readlink project-manager) ]]; then
      _i 'Cannot remove the current generation %s' "$generationId" >&2
    else
      _i 'Removing generation %s' "$generationId"
      "${DRY_RUN_CMD[@]}" rm "${VERBOSE_ARG[@]}" "$linkName"
    fi
  done

  popd > /dev/null || exit
}

function pm_expireGenerations() {
  setProjectManagerPathVariables

  local generations
  generations="$(
    find "$PM_PROFILE_DIR" -name 'project-manager-*-link' -not -newermt "$1" \
      | sed 's/^.*-\([0-9]*\)-link$/\1/'
  )"

  if [[ -n $generations ]]; then
    # shellcheck disable=2086
    pm_removeGenerations $generations
  elif [[ -v VERBOSE ]]; then
    _i "No generations to expire"
  fi
}

function pm_listPackages() {
  setNixProfileCommands
  local outPath
  outPath="$("${LIST_OUTPATH_CMD[@]}" | grep -o '/.*project-manager-path$')"
  if [[ -n $outPath ]]; then
    nix-store -q --references "$outPath" | sed 's/[^-]*-//'
  else
    _i 'No project-manager packages seem to be installed.' >&2
  fi
}

function pm_format() {
  local extraArgs=("$@")

  if [[ -v __PM_FORMATTER ]]; then
    "$__PM_FORMATTER" "${extraArgs[@]}"
  else
    _iWarn 'Not in a Project Manager environment. Attempting `nix fmt`, which is much slower' >&2
    setFlakeAttribute
    nix fmt "${extraArgs[@]}"
  fi
}

function newsReadIdsFile() {
  local dataDir="${XDG_DATA_HOME:-$HOME/.local/share}/project-manager"
  local path="$dataDir/news-read-ids"

  # If the path doesn't exist then we should create it, otherwise
  # Nix will error out when we attempt to use builtins.readFile.
  if [[ ! -f $path ]]; then
    mkdir -p "$dataDir"
    touch "$path"
  fi

  echo "$path"
}

# Builds the Project Manager news data file.
#
# Note, we suppress build output to remove unnecessary verbosity. We
# put the output in the work directory to avoid the risk of an
# unfortunately timed GC removing it.
function pm_buildNews() {
  local newsNixFile="$1"
  local newsJsonFile="$WORK_DIR/news.json"

  # TODO: Use check=false to make it more likely that the build succeeds.
  pm_buildFlake \
    "$FLAKE_CONFIG_URI.config.news.json.output" \
    --quiet \
    --out-link "$newsJsonFile" \
    || return

  local extraArgs=()

  for p in "${EXTRA_NIX_PATH[@]}"; do
    extraArgs=("${extraArgs[@]}" "-I" "$p")
  done

  local readIdsFile
  readIdsFile=$(newsReadIdsFile)

  # nix-instantiate \
  #     --no-build-output --strict \
  #     --eval '<project-manager/project-manager/build-news.nix>' \
  #     --arg newsJsonFile "$newsJsonFile" \
  #     --arg newsReadIdsFile "$readIdsFile" \
  #     "${extraArgs[@]}" \
  #     > "$newsNixFile"
}

function pm_showNews() {
  setWorkDir
  setFlakeAttribute

  local newsNixFile="$WORK_DIR/news.nix"
  pm_buildNews "$newsNixFile"

  local readIdsFile
  readIdsFile=$(newsReadIdsFile)

  local news

  # shellcheck disable=2154,2046
  case $1 in
    --all)
      news="$(nix-instantiate --quiet --eval --expr "(import ${newsNixFile}).news.all")"
      ;;
    --unread)
      news="$(nix-instantiate --quiet --eval --expr "(import ${newsNixFile}).news.unread")"
      ;;
    *)
      _i 'Unknown argument %s' "$1"
      return 1
      ;;
  esac

  # Prints the news without surrounding quotes.
  echo -e "${news:1:-1}" | ${PAGER:-less}

  local allIds
  allIds="$(nix-instantiate --quiet --eval --expr "(import ${newsNixFile}).meta.ids")"
  allIds="${allIds:1:-1}" # Trim surrounding quotes.

  local readIdsFileNew="$WORK_DIR/news-read-ids.new"
  {
    cat "$readIdsFile"
    echo -e "$allIds"
  } | sort | uniq > "$readIdsFileNew"

  mv -f "$readIdsFileNew" "$readIdsFile"
}

function pm_uninstall() {
  pm_setVerboseAndDryRun
  setNixProfileCommands

  _i 'This will remove Project Manager from your system.'

  if [[ -v DRY_RUN ]]; then
    _i 'This is a dry run, nothing will actually be uninstalled.'
  fi

  local confirmation
  read -r -n 1 -p "$(_i 'Really uninstall Project Manager?') [y/n] " confirmation
  echo

  # shellcheck disable=2086
  case $confirmation in
    y | Y)
      _i "Switching to empty Project Manager configuration..."
      PROJECT_MANAGER_CONFIG="$(mktemp --tmpdir project-manager.XXXXXXXXXX)"
      echo "{ lib, ... }: {" > "$PROJECT_MANAGER_CONFIG"
      echo "  project.file = lib.mkForce {};" >> "$PROJECT_MANAGER_CONFIG"
      echo '  project.stateVersion = 0;' >> "$PROJECT_MANAGER_CONFIG"
      echo "  manual.manpages.enable = false;" >> "$PROJECT_MANAGER_CONFIG"
      echo "}" >> "$PROJECT_MANAGER_CONFIG"
      pm_switch
      if [[ -e $PM_PROFILE_DIR ]]; then
        "${DRY_RUN_CMD[@]}" "${REMOVE_CMD}" "$PM_PROFILE_DIR/project-manager" project-manager-path || true
      fi

      rm "$PROJECT_MANAGER_CONFIG"

      if [[ -e $PM_DATA_HOME ]]; then
        "${DRY_RUN_CMD[@]}" rm "${VERBOSE_ARG[@]}" -r "$PM_DATA_HOME"
      fi

      if [[ -e $PM_STATE_DIR ]]; then
        "${DRY_RUN_CMD[@]}" rm "${VERBOSE_ARG[@]}" -r "$PM_STATE_DIR"
      fi

      if [[ -e $PM_PROFILE_DIR ]]; then
        "${DRY_RUN_CMD[@]}" rm "${VERBOSE_ARG[@]}" "$PM_PROFILE_DIR/project-manager"*
      fi
      ;;
    *)
      _i "Yay!"
      exit 0
      ;;
  esac

  _i "Project Manager is uninstalled but your project.nix is left untouched."
}
